### ISSUE TITLE
The big V2 refactoring and API thread

### ISSUE THREAD

**Issue by deanishe** [deanishe/alfred-workflow#55]
A significant refactoring of Alfred-Workflow is long overdue. The `Workflow` class has grown very large and the unittests are a horror show, largely due to the necessity of a workflow-like environment to run them in.

I'd love to hear your ideas on how the library should be refactored and how the API should be changed.

My thoughts so far:
### Refactoring
- Move all the code that requires a workflow environment (`info.plist` or `version` files, Alfred environmental variables) to its own module(s).
- Split the different feature groups of `Workflow` into separate modules (data storage and cache functions, settings, filtering, text processing, serialisation etc.).
- Generalise the update API, so other backends (e.g. RSS, Packal) can be added easily.
- Add lots of mocking to unittests, especially to return canned HTTP responses and avoid hitting web APIs.
### API changes
- Change default `max_age` to 0 in `Workflow.cached_data()`, so cached data is always returned regardless of age by default.
- Move `version` from `update_settings` dict to a `Workflow` argument (as an alternative to `version` file).
- (Possibly) allow `filter()` to be called with an empty `query`, whereupon it will return all `items`. This would mirror the way the built-in Python `filter()` function works, and allow @smargh to not bother checking whether `query` is empty.

Paging @smargh @fniephaus @owenwater :grinning:

**Comment by deanishe** [comment]
So this is the current structure I'm working towards with v2:


- workflow/
    - \_\_init\_\_.py
        - `Workflow` class — Convenience wrapper around functions from other modules
    - background.py
        - Control background processes
    - common.py
        - Logging, decorators
    - constants.py
        - Library constants, e.g. icons, string replacements
    - core.py
        - Alfred-related functionality
            * info.plist parsing
            * Alfred environmental vars
    - storage.py
        - Data storage/caching
    - search.py
        - Indexing and searching
    - settings.py
        - Settings object
    - update.py
        - Update from GitHub
    - version
        - Current version number
    - web.py
        - HTTP

I also want to move to [py.test](http://pytest.org/latest/) for unit test. It's much more powerful and has plugins that mean that it's no longer necessary to make actual HTTP(S) calls or run tests from a wrapper script that sets up the testing environment.

**Comment by fniephaus** [comment]
Hi @deanishe,
Looks great! Good work.

I'm not sure if you really need to move to py.test. Check this out:
http://cramer.io/2014/05/20/mocking-requests-with-responses/

I'm also looking for an easy way to test HTTP requests in my workflows. Maybe this is something `Alfred.Workflow` could also provide? As simple test suite for workflows?

**Comment by deanishe** [comment]
Hey @fniephaus,

Thanks, but I haven't actually done any work yet…

I've looked at mocking responses etc. The problem as I see it is, I am (or we are) going to have to mock out a whole load of situations to properly test web.py. That leaves us with a lot of new code that also needs testing…

py.test has a few awesome plugins that can make all that unnecessary. In particular, there's an httpbin plugin that starts a local instance of httpbin (which is what 90% of the tests use). That would make it super easy to rewrite the tests.

There's a local HTTP server plugin that should cover the rest of web.py and make testing the update code that calls the GitHub API as simple as grabbing the releases JSON with curl and pointing the plugin at it. 

What's more, the powerful fixtures API in py.test should make it easy to create the Alfred-/workflow-like environment a lot of the tests require.

py.test seems to be capable of solving 90% of the issues with the unit tests without us having to write a bunch of support code that in turn needs testing.

I'll add some links later. I'm currently stuck in the RegionalBahn on my phone.

**Comment by deanishe** [comment]
So, what I was looking at was:

- https://pypi.python.org/pypi/pytest-httpbin
- https://pypi.python.org/pypi/pytest-localserver

I can't find any of the other stuff right now, as I'm a wee bit drunk, but there was other good stuff that should make the tests relying on a workflow-like environment (env vars, info.plist etc.) really easy to test.

**Comment by deanishe** [comment]
Also, I think `core.py` is a crappy name. I've thought about `alfred.py`, but that's also a bit crappy.

What's a good name for the module that contains all the Alfred-dependent variables/functions?

**Comment by m1keil** [comment]
workflow.py? :-)

**Comment by deanishe** [comment]
> workflow.py? :-)

It's the obvious answer.

Tbh, I'm only after a different name because I'm sick of seeing `workflow.workflow` in the docs…

**Comment by smargh** [comment]
A la the `info.plist` maybe `info.py`. Or `environ.py` since its information about the workflow environment. Possibly also `details.py` for similar reasons. I personally like the semantics of something like `from details import name, bundle_id`.

**Comment by deanishe** [comment]
@fniephaus @m1keil @owenwater @smargh 
So, I've pushed the alpha of v2 to the [v2 branch](https://github.com/deanishe/alfred-workflow/tree/v2).

**TL;DR**

What about…

- The data caching/storage API?
- How much magic is too much magic?
- A config file. And if it's okay, which format?

#### Current status

Mostly v2 has been about moving code around and moving towards `py.test`.

I've move code from `workflow.py` into a *lot* of different files, but somehow it's still ~1500 lines long :disappointed:

I've been moving towards 3 goals:

- Allow the user to do (almost) everything `Workflow` does without `Workflow`. The `Workflow` class just becomes a higher-level API.
- Zero import side-effects. A problem with v1 is that importing any part of Alfred-Workflow often meant a `Workflow` instance being created. That's mostly fixed (in a very hacky fashion) in the most recent version of 1.x, but it's hacky. Libraries like `update.py` should remain passive as much as possible.
- A hook-based plugin architecture. Currently, there are no plugins and only a rudimentary signal/plugin system in place. In terms of core functionality, I want to move serializers, magic arguments and updating to plugins.

#### Data storage/caching

Currently, I'm stuck at the point of refactoring the data storage/caching API. It's required outside of `Workflow`, so it can't rely on that, but still needs a simple interface. I'm tending towards a `Serializer` superclass that all serializer plugins should inherit from. It would implement the `store(d)_data()` and `cache(d)_data()` APIs, and allow subclasses to only implement the `dump()` and `load()` methods. The superclass would have attributes like `filename`, so that subclasses, like `sqlite`, that can't work with file objects can do their thing without gymnastics.

There'd be a `get_serializer(format)` function in `storage.py`. It would return a object that has all the `stored_data()`, `cached_data()` etc. methods.

So my first question is: Do you think that's a reasonable implementation? Is there an obviously superior one?

#### Workflow variables

Also, I worry if the `env.py` module has a bit too much magic. As is stands, it imports `WorkflowEnvironment` from `_env.py` and essentially replaces itself (the `env` module) with it (the `WorkflowEnvironment` instance) on import. `WorkflowEnvironment` contains all the Alfred environmental variables and the stuff from `info.plist`, too. It loads environmental variables on instantiation, but `info.plist` variables on request (delaying parsing).

So my second question is: Is that too much magic? It's a nice touch from a user perspective, but it makes writing tests a pain in the arse (though the `WorkflowEnv` class in `tests/util.py` solves those issues). Is the unit test hack good enough, or should the use of `WorkflowEnvironment` be explicit? (`Workflow` would hide most of the difference in either case.)

#### `Serializer` and other APIs

So, I think the `Serializer` superclass should stick fairly closely to the existing data storage/caching APIs. What about the `Updater` and `MagicArgument` superclasses? What would be the best API for them to hook into `Workflow` (or the library in general)?

#### Specifying plugins and configuration

What's the best way for AW users to specify plugins? As an argument to `Workflow`? Via an API? Via a configuration file?

@shawnrice and I have been talking about a general configuration file for workflows. He would have his data for Packal and I would have mine for AW configuration.

He's more in favour of JSON, I want INI (I think it's more friendly to non-savvy users and doesn't break when TextEdit is used).

What do you think?

#### Other questions

Is there anything else that could be moved to a signal+plugin architecture?

**Comment by smargh** [comment]
> I've been moving towards 3 goals:
> 
> + Allow the user to do (almost) everything Workflow does without Workflow. The Workflow class just becomes a higher-level API.
> + Zero import side-effects. A problem with v1 is that importing any part of Alfred-Workflow often meant a Workflow instance being created. That's mostly fixed (in a very hacky fashion) in the most recent version of 1.x, but it's hacky. Libraries like update.py should remain passive as much as possible.
> + A hook-based plugin architecture. Currently, there are no plugins and only a rudimentary signal/plugin system in place. In terms of core functionality, I want to move serializers, magic arguments and updating to plugins.

I love the layered approach. Leave a single, simple access point for most users (workflow authors), but allow for others to tweak/build. I think this will also make it easier for changes and upgrades.

I've read the hook code, but don't quite get it. I will need to sit with that for a bit, but I like the overall idea. Again, stronger decoupling makes for easier evolution.


>## Data storage/caching
> 
> Currently, I'm stuck at the point of refactoring the data storage/caching API. It's required outside of Workflow, so it can't rely on that, but still needs a simple interface. I'm tending towards a `Serializer` superclass that all serializer plugins should inherit from. It would implement the `store(d)_data()` and `cache(d)_data()` APIs, and allow subclasses to only implement the `dump()` and `load()` methods. The superclass would have attributes like `filename`, so that subclasses, like `sqlite`, that can't work with file objects can do their thing without gymnastics.
> 
> There'd be a `get_serializer(format)` function in `storage.py`. It would return a object that has all the `stored_data()`, `cached_data()` etc. methods.
> 
> So my first question is: Do you think that's a reasonable implementation? Is there an obviously superior one?
 
This makes good sense to me. One level of class inheritance isn't too gnarly. It keeps things well contained, and allows for evolution (like adding a `sqlite` serializer). My only question is about the user-facing API. I'm assuming the `Workflow` class will function the same (e.g. `Workflow.store_data(data, serializer=name)`). Does making the serializer the class make the API odd or confusing, like `JSON.store_data(data)`? Or, are you planning on making module functions within `storage.py` that mirror the `Workflow` implementation without the class (e.g. `storage.store_data(data, serializer=name)`). I think having `Serializer` objects that define their own `dump()` and `load()` methods makes a lot of sense, but this feels like something that the user shouldn't need to see. They specify a different serializer if they want one as a param on storing, not as the object to do the storing.

>## Workflow variables
> 
> Also, I worry if the `env.py` module has a bit too much magic. As is stands, it imports `WorkflowEnvironment` from `_env.py` and essentially replaces itself (the `env` module) with it (the `WorkflowEnvironment` instance) on import. `WorkflowEnvironment` contains all the Alfred environmental variables and the stuff from info.plist, too. It loads environmental variables on instantiation, but info.plist variables on request (delaying parsing).
> 
> So my second question is: Is that too much magic? It's a nice touch from a user perspective, but it makes writing tests a pain in the arse (though the WorkflowEnv class in tests/util.py solves those issues). Is the unit test hack good enough, or should the use of WorkflowEnvironment be explicit? (Workflow would hide most of the difference in either case.)

Again, I've read the code but don't fully understand what's happening. Why can't you do all of this in `env.py`? Could you not put the `WorkflowEnvironment` class in `env.py`, and then put
```
e = WorkflowEnvironment()
sys.modules[__name__] = e
```
at the bottom of the module? Why put it in a separate file? Or, why not just create aliases in the `env.py` module namespace, like:
```
e = WorkflowEnvironment()
info = e.info
logfile = e.logfile
#etc...
```
Not as clean perhaps, but totally understandable as to what's happening.

But all of these thoughts are predicated on me not knowing what exactly is happening, so there may very well be implementation details that makes such things impossible. Just thoughts and suggestions.

>## Serializer and other APIs
> 
> So, I think the `Serializer` superclass should stick fairly closely to the existing data storage/caching APIs. What about the `Updater` and `MagicArgument` superclasses? What would be the best API for them to hook into `Workflow` (or the library in general)?

Need to think on this.

>## Specifying plugins and configuration
> 
> What's the best way for AW users to specify plugins? As an argument to Workflow? Via an API? Via a configuration file?
> 
> @shawnrice and I have been talking about a general configuration file for workflows. He would have his data for Packal and I would have mine for AW configuration.
> 
> He's more in favour of JSON, I want INI (I think it's more friendly to non-savvy users and doesn't break when TextEdit is used).
> 
> What do you think?

JSON allows for robust data types, but does break easily. If you want something "unbreakable", then I think striving for `key=value` string to string configs (a la Bash) is probably safest.

- - -

All in all, things look good for an alpha. I think moving to a layered setup, with flexibility but sane defaults is great. The best but trickiest part of `Alfred-Workflow` is that it is used by people who basically know no Python all the way to professional Python developers (and oddballs like me in the middle). So, aiming to have one thing that works awesomely for both groups is undoubtedly difficult, but this looks like it is 100% on the right path.

**Comment by owenwater** [comment]
### Data storage/caching

It's definitely a good implmentation. But I don't think `Serializer` is necessary. For example, the attribute like 'filename' as you mentioned, a serializer may not need it at all, e.g. a serializer that depends on multiple files. I just feel like duck type is simple and pythonic.

### Workflow variables

It took me some time to fully understand your purpose here and the complication here is absolutely necessary. A good feature should not be removed just because of the unit test. 

### Plugin APIs

I'm not fully understood `hook.py` here. But based on my understanding, a plugin contains some features that may not be needed by everyone so it should be easy to develop/read/install/use. My idea is based on two different types of plugins:

- Enhancement. Adding new features that don't affect the existing code.
- Replacement. Changing the existing code.

There aren't much things needed to be done for the first one. For the second one, a plugin loader should be able to replace the function automatically who shares the same name.

For example, to implement `pretty_xml` plugin, all we need to do is to create a new `workflow.py/Workflow.send_feedback()` function just like the one in main code but this one is actully in the plugins folder.

### Configuration

I think a config file for plugins is too much. People should be able to enable a plugin by throwing a file/directory into `plugins/` which just like vim or zsh.

### Other things

There are some classes inherit the `dict` class but haven't overriden some functions like `clear()`, `pop()`, `popitem()`.

**Comment by deanishe** [comment]
@smargh

**Tip:** I keep editing your comments to add this, so … When you have a code block, append the language name to the opening ticks to get syntax highlighting:

    ```python
    # Call external processes
    from subprocess import Popen
    debug = True
    ```

gives you:

```python
# Call external processes
from subprocess import Popen
debug = True
```

> Could you not put the `WorkflowEnvironment` class in `env.py`

Not without a lot of messing around. Once the `WorkflowEnvironment` instance is put in `sys.modules` it can no longer see anything in the `env.py` namespace, e.g. `os`, `plistlib` or `util`, so method calls fail. Initially, I aliased all the other globals within `WorkflowEnvironment.__init__`, but in the end it was cleaner to put `WorkflowEnvironment` in a separate module and define it normally.

> Or, why not just create aliases in the `env.py` module namespace

`WorkflowEnvironment` is lazy-loading. It won't explode without an `info.plist` until you try to access one of its variables that it loads from `info.plist`. If you start pulling things like `info` and `logfile` out of it, you won't be able to import `env.py` without an `info.plist` present.

As I said, I want to avoid import side-effects as much as possible. It makes testing much easier in particular, and is bad form in general. `env.py` already crosses that line (which is why I asked if it's "too much magic"), but it only fiddles with itself and won't immediately explode if it's not in a workflow environment.

The alternative is to use the `WorkflowEnvironment` class directly:

```python
from env import WorkflowEnvironment
env = WorkflowEnvironment()
...
```

or remove a lot of the magic and add an instance to `env.py`:

In `env.py`:

```python
class WorkflowEnvironment(object):
    ...

env = WorkflowEnvironment()
```

Then in other files:

```python
from env import env
...
```

> My only question is about the user-facing API. I'm assuming the `Workflow` class will function the same (e.g. `Workflow.store_data(data, serializer=name)`)

Yup. The `Workflow` API should remain largely the same. It's just a question of what the best implementation is. There will be some changes, though, which will be backwards-incompatible in many cases.

For example, I want to remove `Workflow.alfred_env` and `Workflow.info` and replace them with `Workflow.env` (which is an alias for the `workflow.env` module), which combines them. The contents of `info.plist` would still be available at `Workflow.env.info`.

@owenmin
> But I don't think `Serializer` is necessary. For example, the attribute like 'filename' as you mentioned, a serializer may not need it at all, e.g. a serializer that depends on multiple files. I just feel like duck type is simple and pythonic.

Concrete serializers wouldn't be required to inherit from `Serializer`, they'd just need `load()` and `dump()` methods as they do now.

The purpose is more to put the caching/storage logic into the `Serializer` class where concrete serializers can override or change it if necessary, so they aren't limited to the `load()` + `dump()` API.

Currently, serializers get an object and an open file handle to work with. The `sqlite`-based serializer we're working on needs a filename, however. I think it'd be a cleaner solution if it could hook into the `Serializer` superclass to grab the filename instead of trying to extract it from an open file handle.

This would be an implementation detail and wouldn't affect the API.

> I'm not fully understood `hook.py` here

Not surprising, as it doesn't actually do anything yet. The plugin system is modelled on [pelican](https://github.com/getpelican/pelican)'s. It will (hopefully) work like this:

1. Add directories containing plugins with `plugins.add_plugin_path('some/dir/path')`.
2. Load named plugins with `plugins.load_plugins(['plugin1', 'plugin2', ...])`.
3. Import each module whose name was specified and call its `register()` function. Typically, `register()` will import `hooks.py` and subscribe (connect) to one or more of the signals.
4. `Workflow` (mostly) sends signals when something happens (e.g. initialisation) or it wants something (e.g. a serializer for a given format).
5. Plugins respond to the signals.

I'm undecided whether all plugins should be loaded automatically or whether they need to be explicitly activated.

I'm not sure if plugins will prove to be of any use to users, but I think using them makes (or forces me to make) the internal API cleaner.

> People should be able to enable a plugin by throwing a file/directory into plugins/ which just like vim or zsh.

I was thinking more as a way to *disable* plugins, e.g. turn off magic arguments. Mostly, I see it as an alternative to the arguments to `Workflow`. If you have lots of Script Filters and a whole bunch of arguments to `Workflow`, being able to put them in a `workflow.ini` file might be easier. DRY and all that. It would be entirely optional.

It would also be a handy place to turn debugging on/off and add any settings your own plugins or scripts need:

```ini
[default]
version=1.1

[plugins]
; Add custom plugin path
paths=./plugins

[[enabled]]
; Your custom plugins
updater_bitbucket
serializer_xls
serializer_tsv
serializer_csv

[[disabled]]
; Turn off magic arguments
magic_default

[update]
; Settings for the updater
bitbucket_repo=deanishe/alfred-convert
; Check for new version every week
interval=7

[xls]
; Custom settings understood by your plugin(s)
format=xlsx
```

> a plugin loader should be able to replace the function automatically who shares the same name

That's the idea. The default plugin directory would be last in the resolution order, so a custom plugin with the same name would override a built-in one.

> For example, to implement `pretty_xml` plugin

There's already a stub of one in the `plugins/` directory and I've added an `xml_generator_done` signal that is called with the generated XML.

`feedback.XMLGenerator` could easily be discovered via hooks, though, if being able to use a custom generator were useful.

> There are some classes inherit the `dict` class but haven't overriden some functions like `clear()`, `pop()`, `popitem()`

Well spotted. `PersistentDict` definitely needs a `clear()` method. I'm not sure if it's a good idea to add the other methods, though, especially to `WorkflowEnvironment`.

**Comment by owenwater** [comment]
Thanks for the explanation

### Data storage
That makes more sense to me. It's good to have a superclass if multiple serializers share code.

### Hook

My original idea which might not be very clear in the last post was using reflection but not hook. Because a hook architecture requires implmentation for every place that can be injected. However, it turned out not a bad idea after I read your example and thought it again this morning. Because the hook looks much safer and it's enough for internal usage.

### Config file
A configuration file is great if it's not only includes enabled/disabled plugins but also many other workflow related information. The only concern I got here is having both of `[[enable]]` and `[[disable]]`.

### Dictionary
The `pop()` and `popitem()` is important for `PersistenDict` because it's a low level class. User should be able to use it just as a dictionary without any suprise. Because those two funcitons can change the content of a dictionary, the `save()` function should be called after.
The `WorkflowEnvironment` is a read-only class for me. My suggestion is throwing an exception for any setting attempt.

**Comment by deanishe** [comment]
> My original idea which might not be very clear in the last post was using reflection but not hook

That was my original plan, before having a look at a few plugin/hook systems. The thing I like about the planned implementation is the ability to add hooks just about anywhere.

> The only concern I got here is having both of `[[enable]]` and `[[disable]]`

I agree. As I said, I'm unsure whether to enable all plugins by default or require them to be explicitly enabled. I suppose on-by-default and lose the `[[enabled]]` section.

> The `pop()` and `popitem()` is important for `PersistentDict` because it's a low level class

I've never used them, but I've added them.

**Comment by owenwater** [comment]
I'll vote for `[[disable]]` section as well since it's simple for most people.

**Comment by shawnrice** [comment]
I'll concede the json to ini point. The only resistance that I ever had was that I've always thought that the ini extension is ugly (based on previous windows experience), that ini files themselves look ugly (based on other aesthetic aspects), that json files, for some strange reason, look pretty.

But all the reasons that that make JSON fail are compelling to switch to INI, and INI is easy enough to implement in PHP (and I'm assuming Ruby as well).

We just have to decide on a canonical filename. While `info.ini` would make sense for reasons of symmetry, we probably should go with `workflow.ini` because that just makes more sense.

I'm also planning on implementing this, somehow, in a PHP library that I'm writing.

It would be nice to have the `ini` file sectioned in a way that would indicate "global", "packal", "alfred workflow", "alphred", etc.... but I'm not sure if that will provide too many barriers on the user side.

**Comment by deanishe** [comment]
INI *is* ugly and JSON *is* far prettier, but INI is much easier for humans to edit, is natively supported by at least PHP and Python, and — importantly — supports comments (not allowed in canonical JSON).

We'd have to figure out a useable syntax subset supported by PHP and Python: they appear to use different syntaxes for nested sections.

Hadn't given the name much thought, tbh. `workflow.ini` would be a good choice for the reasons you give.

> It would be nice to have the ini file sectioned in a way that would indicate "global", "packal", "alfred workflow", "alphred", etc...

Indeed. There's common stuff, though. We'll have to figure out which variables belong in a `default` or generic `workflow` section (version?)

**Comment by OwenMin** [comment]
Can you guys stop including me in every email you send to each other? 

Thank you.

**Comment by shawnrice** [comment]
You need to unsubscribe from the GH issue. The emails are automatic.

**Comment by deanishe** [comment]
:grinning: 

GitHub *does* send far too many emails, though. I had to unfollow most of the repos I was following on account of the silly amount of emails. Even the ones I've contributed to.

**Comment by deanishe** [comment]
Oh wow. He's deleted his account.

I hope that wasn't all "our" emails…

**Comment by smargh** [comment]
How strange. He never contributed a pull request or anything, did he? He simply started "watching" the repo? That sucks.

**Comment by deanishe** [comment]
My mistake. He didn't delete his account. I just got his name (Owen Min) confused with his username (owenwater).

Yeah, he contributed [2 pull requests](https://github.com/deanishe/alfred-workflow/pulls?q=is%3Apr+author%3Aowenwater+is%3Aclosed). Some very valuable input, too. I really hope he chips in again in the future.

**Comment by smargh** [comment]
Ah. I see. Yeah, those pull requests were quite good (better than all of mine). And you'll always meet some unique people on the interwebs. It gets alot more likely when you have a popular project (like this one).

**Comment by deanishe** [comment]
To be fair, we discussed his accepted PR in detail in #42 before he submitted it (although I wish I'd accepted his other PR instead of implementing it myself and then having to add his additional — and very good — code by hand later).

The issue with your pull requests (and perhaps more generally), as @shawnrice and I have noted, is you tend to get too caught up in the specific nut you're trying to crack. As a result, you sometimes overlook how your idea might break other stuff (e.g. #51). It's always a good idea to ask yourself (a) why is what I'm planning bad/wrong and (b) can I generalise this to make it more useful?

**Comment by owenwater** [comment]
Github does generate lots of email, feel sorry for spam that poor guy so much. However, that's still interesting to have someone's id same as my name.

BTW, I can write some code this weekend. Do you get anything has high priority but haven't started yet for v2 so I can help? Just don't want to make the same mistake again. :smile:

**Comment by deanishe** [comment]
I'm confused as fuck now. Who was it complained again? I'd got it into my head that it was you @owenwater, but I'm a bit drunk right now.

Oh wait. Did one of us type @owenmin instead of @owenwater? Haha!

Regarding coding, nothing that still needs implementing has a fixed API yet. As a result, I'm hesitant to ask someone to have a go at writing it for fear that it will need rewriting.

I'm currently working on the cached/stored data API, but all I've done so far is cut-and-paste it into `storage.py` (I haven't actually run any of the code yet).

The update system needs refactoring around hooks/plugins, but depends on the caching API. Keychain and magic args need moving out of `Workflow`, but they're mostly cut-and-paste jobs.

I think what would help most at this point would be an "audit" of the code I've written/moved around so far, and some brainy thinking about what's wrong, what's missing and how the update, serializer and magic argument APIs should work.

- Have I done anything obviously bone-headed (code and tests)?
- Is there anything that could be refactored/generalised (code and tests)?
- Are there other hooks that could/should be added?
- Assuming hooks can be used to override built-in functions, which functions should be made overrideable? XML generation? `Workflow.filter()`?
- What should you be able to specify in `workflow.ini` apart from version and update settings? How should it be integrated with `Workflow`? Does it go in `Workflow` or `env`? (I'm thinking `env`, so plugins can access the configuration without having to instantiate `Workflow`, which is something I want to avoid.)
- What else could/should be added to `env`? For example, I've been thinking about `theme_light`/`theme_dark`, which would be a boolean calculated from the colour of the theme background, so users would know whether to use light or dark icons.
- **How should be the update API work?**

**Comment by shawnrice** [comment]
If you want the theme light/theme dark, then just steal the algorithms that I wrote for the bundler. All the color math is already there, and the research is done.

**Comment by deanishe** [comment]
> just steal the algorithms that I wrote for the bundler

Goes without saying, tbh.

**Comment by shawnrice** [comment]
I just wrote that in there in case someone else decided to tackle it and didn't know about the color work done in the bundler.

But, for ease, here is the consolidated code from the PHP bundler. The key calculation is the one that sets the luminance variable.

````php
  function light_or_dark() {
  	// Regex pattern to parse the Alfred background variable
    $pattern = "/rgba\(([0-9]{1,3}),([0-9]{1,3}),([0-9]{1,3}),([0-9.]{4,})\)/";
    // Do the regex, matching everything in the $matches variable
    preg_match_all( $pattern, $_ENV[ 'alfred_theme_background' ], $matches );
    // Pull the values into an $rgb array
    $rgb = array( 'r' => $matches[1][0], 'g' => $matches[2][0], 'b' => $matches[3][0] );

    // This calculates the luminance. Values are between 0 and 1.
    $luminance = ( 0.299 * $rgb[ 'r' ] + 0.587 * $rgb[ 'g' ] + 0.114 * $rgb[ 'b' ] ) / 255;

    if ( 0.5 < $luminance ) {
        return 'light';
    }
    return 'dark';
  }
````

**Comment by fniephaus** [comment]
I use this:
```python
def is_dark(wf):
      return min([int(x) for x in wf.alfred_env['theme_background'][5:-6].split(',')]) < 128
```

**Comment by deanishe** [comment]
How has that worked out for you @fniephaus? @shawnrice says the luminance is the critical calculation, but that bits missing from yours. Does it matter?

I have no idea myself.

Also, frohes Neues :fireworks:

**Comment by fniephaus** [comment]
I haven't really given it that much thought, so Shawn's algorithm is probably more accurate.
Nonetheless, it does what it should be doing except maybe for weird colors that I haven't tested...

**Comment by shawnrice** [comment]
The luminance calculates the "brightness" of the image. The human eye sees green light really well and blue light terribly. So, the coefficients weight that appropriately.

Granted, any difference is probably going to be an edge case.

**Comment by fniephaus** [comment]
Just for the record, here's a Python equivalent of @shawnrice's algorithm:
```python
def is_dark():
    rgb = [int(x) for x in WF.alfred_env['theme_background'][5:-6].split(',')]
    return (0.299 * rgb[0] + 0.587 * rgb[1] + 0.114 * rgb[2] ) / 255 < 0.5
```

**Comment by idpaterson** [comment]
@deanishe have you had any involvement in testing Alfred 3.0? [It looks like there are a lot of interesting changes coming](https://www.alfredapp.com/blog/announcements/alfred-v3-is-coming/) though the details will be shared in follow-up blog posts over the next few weeks. It is not clear yet to what extent the changes to workflows would affect workflow libraries like alfred-workflow.

**Comment by deanishe** [comment]
@idpaterson I know nothing about Alfred 3 beyond a few educated guesses based on what Andrew has expressed interest in on the forums. Everything in that blog post is news to me.

**Comment by shawnrice** [comment]
From my discussions with Andrew, most Alfred 2 workflows should still be compatible with Alfred 3. There will definitely be improvements, but I'm not sure either how much the libraries will need to change or what new stuff should be added in so as to take advantage of newer features.

**Comment by idpaterson** [comment]
That is reassuring. From the screenshot it looks like primarily changes within the info.plist but I'm looking forward to learning more as it is unveiled on the blog over the coming weeks.

**Comment by deanishe** [comment]
Closing this thread. V2 Is coming, and I’m going to start a more up-to-date issue.
